pykwasm/src/pykwasm/kdist/wasm-semantics/ulm-wasm.md

```diff
@@ -1,21 +1,120 @@
+
+The ULM-integrated WASM VM has two possible program encodings:
+
+1.  In the local test VM case, a direct encoding is used.
+
+    ```local
+    syntax PgmEncoding ::= ModuleDecl
+    ```
+
+2.  In the remote ULM-integrated VM case, a byte encoding is used.
+
+    ```remote
+    syntax PgmEncoding ::= Bytes
+    ```
+
+```k
 endmodule
 ```
 
 ```k
 module ULM-WASM
-  imports WASM-TEXT
-  imports ULM
+    imports ULM-WASM-SYNTAX
+    imports ULM
+```
 
-  configuration <ulm-wasm>
-                  <wasm/>
-                </ulm-wasm>
+Program Decoding
+----------------
+
+The WASM VM must decode the input program:
+
+1.  In local test VM case, the decoding function is just identity.
+
+    ```local
+    syntax ModuleDecl ::= decodePgm(ModuleDecl) [function, total]
+    // ----------------------------------------------------------
+    rule decodePgm(Mod) => Mod
+    ```
+
+2.  In the remote ULM-integrated VM case, a specialized, hooked byte decoder is used.
+
+    ```remote
+    syntax ModuleDecl ::= decodePgm(Bytes) [function, hook(ULM.decode)]
+    ```
+
+Configuration
+-------------
+
+Here we define the initial configuration.
+Note that the default status code indicates an internal error; this is used defensively, since if we ever get stuck, an error will always be indicated.
+Similarly, we define a default null output which may indicate internal errors.
+
+```k
+    syntax OutputData ::= "NO_OUTPUT"
+                        | Bytes
+
+    configuration
+      <ulmWasm>
+        <k> $PGM:PgmEncoding </k>
+        <wasm/>
+        <create> $CREATE:Bool </create>
+        <entry> $WASM:String </entry>
+        <gas> $GAS:Int </gas>
+        <status> EVMC_INTERNAL_ERROR </status>
+        <output> NO_OUTPUT </output>
+      </ulmWasm>
+```
+
+Passing Control
+---------------
+
+The embedder loads the module to be executed and then resolves the entrypoint function.
+
+```k
+    rule <k> PGM:PgmEncoding => . </k>
+         <entry> FUNCNAME </entry>
+         <instrs> .K => decodePgm(PGM) </instrs>
+```
+
+ULM Hook Behavior
+-----------------
+
+These rules define various integration points between the ULM and our Wasm interpreter.
+
+**Note**: the first three rules hooks below are written with helper functions
+          because parse errors were encountered when writing `<generatedTopCell>` literals.
+
+```k
+    syntax Int ::= #getGasLeft() [function, total]
+    rule [[ #getGasLeft() => Gas ]]
+         <gas> Gas:Int </gas>
+
+    syntax Bytes ::= #getOutput() [function, total]
+    rule [[ #getOutput() => #if OutVal ==K NO_OUTPUT #then .Bytes #else {OutVal}:>Bytes #fi ]]
+         <output> OutVal:OutputData </output>
+
+    syntax Int ::= #getStatus() [function, total]
+    rule [[ #getStatus() => #if OutVal ==K NO_OUTPUT #then EVMC_INTERNAL_ERROR #else Status #fi ]]
+         <output> OutVal:OutputData </output>
+         <status> Status:Int </status>
+
+    rule getGasLeft(_) => #getGasLeft()
+    rule getOutput(_)  => #getOutput()
```

Comment:
I'm not convinced that this will actually work in ULM, because ULM constructs a `getOutput(configuration)` term and evaluates it. IIRC, if you have a function `f(Something)` that uses the configuration (directly or indirectly) with the `[[ ... ]]` syntax, then kompile transforms it into `f(Something, GeneratedTopCell)` and silently passes the configuration as an argument. I.e., in the case above, the `getOutput` function is actually compiled to `getOutput(GeneratedTopCell, GeneratedTopCell)`, which is not what ULM expects.

I might be wrong here, @traiansf for confirmation.

Also see:
https://github.com/Pi-Squared-Inc/ulm/blob/main/kllvm/ulm_kllvm.cpp#L192-L197

Reply:
As discussed on Slack, it's fine to fix this in a follow-up PR.